### PR TITLE
Move cmd/controller-manager to k8s.io/controller-manager

### PR DESCRIPTION
**Pull request by cici37**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

 /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR is to address the concern: https://github.com/kubernetes/kubernetes/pull/90976#issuecomment-635546451
Moved cmd/controller-manager to k8s.io/controller-manager
Moved GenericControllerManagerConfiguration in k8s.io/controller-manager. 
Moved CloudControllerManagerConfiguration, KubeCloudSharedConfiguration and ServiceControllerConfiguration to k8s.io/cloud-provider.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by cici37** [comment]
cc @cheftako @andrewsykim This pr is for decoupling types between CCM and KCM. And it will also make it ready to move cmd/controller-manager out-tree if we plan to do so.

File: staging/src/k8s.io/cloud-provider/api/types.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	componentbaseconfig "k8s.io/component-base/config"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GenericControllerManagerConfiguration holds configuration for a generic controller-manager
+type GenericControllerManagerConfiguration struct {
```
Comment (cheftako) [inline staging/src/k8s.io/cloud-provider/api/types.go:27]: We want this shared between the KCM and CCM. I don't think we want to decouple this.

**Comment by cici37** [comment]
Will have the PR ready after new k8s.io/controller-manager repo is in place.

**Comment by cici37** [comment]
cc @cheftako @andrewsykim

File: staging/src/k8s.io/controller-manager/app/options/types.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+*/
+
+package options
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	componentbaseconfig "k8s.io/component-base/config"
+)
+
+// ServiceControllerConfiguration contains elements describing ServiceController.
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/app/options/types.go:24]: Can we put the controller configs in sub-directories. Separating controller concerns from controller-manager concerns seems prudent. It seems fine to have service controller here, while its jsut service controller. However when it becomes, service, cloud, cloud lifecycle, route, x, y and z controllers, I think we will want them each in their own sub directory.

File: pkg/controller/service/config/OWNERS [hunk 3]
```diff
@@ -1,14 +0,0 @@
-approvers:
-- api-approvers
```
Comment (cheftako) [inline pkg/controller/service/config/OWNERS:2]: What about https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/config/v1alpha1/zz_generated. deepcopy.go? Either it should also be deleted or we should retain OWNERS.

File: staging/src/k8s.io/controller-manager/app/options/v1alpha1/conversion.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/app/options/v1alpha1/conversion.go:2]: 2020

File: pkg/controller/service/config/OWNERS (hunk 3 above)
Comment (cici37) [reply pkg/controller/service/config/OWNERS:2]: `https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/config/v1alpha1/zz_generated.deepcopy.go` is deleted together with OWNERS file in this PR.

File: staging/src/k8s.io/controller-manager/app/options/types.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (cici37) [inline staging/src/k8s.io/controller-manager/app/options/types.go:2]: @cheftako Should here be 2019 or 2020? Thanks

File: pkg/controller/service/config/OWNERS (hunk 3 above)
Comment (cheftako) [reply pkg/controller/service/config/OWNERS:2]: Sorry somehow missed it when I looked last time. Not sure how I missed it.

File: staging/src/k8s.io/controller-manager/app/options/types.go (hunk 5 above)
Comment (cheftako) [reply staging/src/k8s.io/controller-manager/app/options/types.go:2]: Old files should be left with their original Copyright notice. New files are generally generated with the current year.

**Comment by cici37** [comment]
Wait for PR [92399](https://github.com/kubernetes/kubernetes/pull/92399) to be merged first

**Comment by cici37** [comment]
I have combined changes of [92399](https://github.com/kubernetes/kubernetes/pull/92399) here.

File: cmd/cloud-controller-manager/app/apis/config/types.go [hunk 6]
```diff
@@ -18,8 +18,8 @@ package config
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
-	serviceconfig "k8s.io/kubernetes/pkg/controller/service/config"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline cmd/cloud-controller-manager/app/apis/config/types.go:21]: controllermanagerconfig or cmconfig

File: cmd/kube-controller-manager/app/options/options_test.go [hunk 7]
```diff
@@ -24,13 +24,14 @@ import (
...
 
 	"github.com/spf13/pflag"
-
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/diff"
 	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 	componentbaseconfig "k8s.io/component-base/config"
 	"k8s.io/component-base/metrics"
-	cmoptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline cmd/kube-controller-manager/app/options/options_test.go:32]: same here, cmconfig or controllermanagerconfig

File: staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go [hunk 8]
```diff
@@ -18,8 +18,9 @@ package v1alpha1
 
 import (
 	"k8s.io/apimachinery/pkg/conversion"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:21]: cmconfig or controllermanagerconfig

File: staging/src/k8s.io/controller-manager/options/cloudprovider.go [hunk 9]
```diff
@@ -18,13 +18,12 @@ package options
 
 import (
 	"github.com/spf13/pflag"
-
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline staging/src/k8s.io/controller-manager/options/cloudprovider.go:21]: cmconfig or controllermanagerconfig

File: staging/src/k8s.io/controller-manager/options/generic.go [hunk 10]
```diff
@@ -23,19 +23,19 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	cliflag "k8s.io/component-base/cli/flag"
 	"k8s.io/component-base/config/options"
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline staging/src/k8s.io/controller-manager/options/generic.go:26]: cmconfig or controllermanagerconfig

File: staging/src/k8s.io/controller-manager/options/kubecloudshared.go [hunk 11]
```diff
@@ -18,25 +18,24 @@ package options
 
 import (
 	"github.com/spf13/pflag"
-
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
+	config2 "k8s.io/controller-manager/config"
```
Comment (andrewsykim) [inline staging/src/k8s.io/controller-manager/options/kubecloudshared.go:21]: cmconfig or controllermanagerconfig

File: cmd/cloud-controller-manager/app/apis/config/types.go (hunk 6 above)
Comment (cici37) [reply cmd/cloud-controller-manager/app/apis/config/types.go]: Updated

**Comment by nikhita** [comment]
publishing-bot job is green :tada:

**Comment by cici37** [comment]
@nikhita Thanksssss so much for helping out on this test :)

File: pkg/controller/apis/config/types.go [hunk 12]
```diff
@@ -18,7 +18,8 @@ package config
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	componentbaseconfig "k8s.io/component-base/config"
+	config2 "k8s.io/controller-manager/config"
```
Comment (cheftako) [inline pkg/controller/apis/config/types.go:21]: Can we make this cmconfig rather than config2? We may also want to alias service/config as serviceconfig. If we did that then controller-managers config could just be config. As the base config that would seem to make more sense.

Comment (cici37) [reply pkg/controller/apis/config/types.go]: Updated to use `cmconfig` for /control-manager/config and `serviceconfig` for /service/config

File: cmd/cloud-controller-manager/app/options/options_test.go [hunk 13]
```diff
@@ -23,22 +23,21 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/diff"
 	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 	componentbaseconfig "k8s.io/component-base/config"
-	cmoptions "k8s.io/kubernetes/cmd/controller-manager/app/options"
-	kubectrlmgrconfig "k8s.io/kubernetes/pkg/controller/apis/config"
-	serviceconfig "k8s.io/kubernetes/pkg/controller/service/config"
+	cmconfig "k8s.io/controller-manager/config"
+	cmoptions "k8s.io/controller-manager/options"
+	"k8s.io/controller-manager/service/config"
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/options/options_test.go:32]: Can we keep this as serviceconfig?

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go [hunk 14]
```diff
@@ -21,8 +21,8 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	serviceconfigv1alpha1 "k8s.io/controller-manager/config/v1alpha1"
```
Comment (cheftako) [inline cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go:24]: This should probably become cmconfigv1alpha1 or ctrlmgrconfigv1alpha1.

File: staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go [hunk 15]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2018 The Kubernetes Authors.
+Copyright 2020 The Kubernetes Authors.
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:2]: New files should use the current year. I believe old files (including those being moved) should be left with their original year.

File: staging/src/k8s.io/controller-manager/service/config/doc.go [hunk 16]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/service/config/doc.go:2]: New file 2020.

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go (hunk 14 above)
Comment (andrewsykim) [reply cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go]: +1

**Review by andrewsykim — COMMENTED** [review]
LGTM, have mostly the same minor comments as @cheftako

**Comment by lavalamp** [comment]
it looks like this was approved before freeze? is that right? Should we put it in the milestone?

**Comment by cheftako** [comment]
> it looks like this was approved before freeze? is that right? Should we put it in the milestone?

Will that be sufficient?

**Comment by lavalamp** [comment]
Wait, my bad, it looks like this didn't actually have an approved label? :(

**Review by andrewsykim — COMMENTED** [review]
Can we file an exception for this? Seems pretty low risk and high benefit for external cloud providers consuming cmd/cloud-controller-manager today.

**Comment by lavalamp** [comment]
We can certainly try! If it's low risk and blocking something important, I
think you can make a case? I didn't realize it was blocking something, or I
would have suggested it.

**Comment by nilo19** [comment]
what's blocking it? really need this to get merged.

**Comment by cici37** [comment]
Thanks @cheftako to help on communicating with release manager to see if we could get it to 1.19. Still in progress

**Comment by dims** [comment]
/assign @smarterclayton @liggitt 

Clayton, Jordan,
Needs root approval

**Comment by cici37** [comment]
Gentle reminder :) @smarterclayton @liggitt

File: staging/src/k8s.io/controller-manager/service/config/BUILD [hunk 17]
```diff
@@ -7,7 +7,8 @@ go_library(
         "types.go",
         "zz_generated.deepcopy.go",
     ],
-    importpath = "k8s.io/kubernetes/pkg/controller/service/config",
+    importmap = "k8s.io/kubernetes/vendor/k8s.io/controller-manager/service/config",
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/service/config/BUILD:10]: service/config used to have things like an OWNERS file. Is there a reason we seem to have lost that?

Comment (cici37) [reply staging/src/k8s.io/controller-manager/service/config/BUILD:10]: Thanks for the catch. Updated^

File: staging/src/k8s.io/controller-manager/service/config/doc.go [hunk 18]
```diff
@@ -1,7 +1,5 @@
-// +build !ignore_autogenerated
```
Comment (cheftako) [inline staging/src/k8s.io/controller-manager/service/config/doc.go:1]: This move seems wrong. I think we'll end up with the wrong file history.

Comment (cici37) [reply staging/src/k8s.io/controller-manager/service/config/doc.go:1]: Thanks. Updated to remove this file history

File: staging/publishing/rules.yaml [hunk 19]
```diff
@@ -31,7 +31,7 @@ rules:
       branch: release-1.19
       dir: staging/src/k8s.io/code-generator
     name: release-1.19
-    go: 1.15.2
+    go: 1.15.2.2
```
Comment (cheftako) [inline staging/publishing/rules.yaml:34]: Can we undo this change?

File: staging/publishing/rules.yaml [hunk 20]
```diff
@@ -59,7 +59,7 @@ rules:
       branch: release-1.19
       dir: staging/src/k8s.io/apimachinery
     name: release-1.19
-    go: 1.15.2
+    go: 1.15.2.2
```
Comment (cheftako) [inline staging/publishing/rules.yaml:62]: Ditto

File: staging/publishing/rules.yaml (hunk 19 above)
Comment (cici37) [reply staging/publishing/rules.yaml]: Updated

File: cmd/cloud-controller-manager/app/apis/config/types.go [hunk 21]
```diff
@@ -29,10 +29,10 @@ type CloudControllerManagerConfiguration struct {
 	metav1.TypeMeta
 
 	// Generic holds configuration for a generic controller-manager
-	Generic kubectrlmgrconfig.GenericControllerManagerConfiguration
+	Generic cmconfig.GenericControllerManagerConfiguration
 	// KubeCloudSharedConfiguration holds configuration for shared related features
 	// both in cloud controller manager and kube-controller manager.
-	KubeCloudShared kubectrlmgrconfig.KubeCloudSharedConfiguration
+	KubeCloudShared cmconfig.KubeCloudSharedConfiguration
```
Comment (liggitt) [inline cmd/cloud-controller-manager/app/apis/config/types.go:35]: GenericControllerManagerConfiguration in k8s.io/controller-manager makes sense

KubeCloudSharedConfiguration and ServiceController types should go to k8s.io/cloud-provider, to avoid putting cloud-specific and kubernetes-specific stuff into the generic controller manager repo. KCM can depend on k8s.io/cloud-provider until it stops running cloud loops.

Comment (cici37) [reply cmd/cloud-controller-manager/app/apis/config/types.go]: Updated the PR to move KubeCloudSharedConfiguration and ServiceController types to k8s.io/cloud-provider. Please have a look when you got time😊

File: staging/src/k8s.io/controller-manager/go.mod [hunk 22]
```diff
@@ -5,16 +5,24 @@ module k8s.io/controller-manager
...
 require (
+	github.com/spf13/pflag v1.0.5
+	github.com/stretchr/testify v1.4.0
 	k8s.io/apimachinery v0.0.0
 	k8s.io/apiserver v0.0.0
+	k8s.io/client-go v0.0.0
+	k8s.io/cloud-provider v0.0.0
 	k8s.io/component-base v0.0.0
+	k8s.io/klog/v2 v2.2.0
+	k8s.io/kube-controller-manager v0.0.0
```
Comment (liggitt) [inline staging/src/k8s.io/controller-manager/go.mod:16]: these two dependencies should not exist... I expected cloud-provider and kube-controller-manager to depend on the generic controller-manager module, not the other way around

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go [hunk 23]
```diff
@@ -47,5 +47,5 @@ func SetDefaults_CloudControllerManagerConfiguration(obj *CloudControllerManager
 	// Use the default RecommendedDefaultGenericControllerManagerConfiguration options
 	kubectrlmgrconfigv1alpha1.RecommendedDefaultGenericControllerManagerConfiguration(&obj.Generic)
 	// Use the default RecommendedDefaultServiceControllerConfiguration options
-	serviceconfigv1alpha1.RecommendedDefaultServiceControllerConfiguration(&obj.ServiceController)
+	cmconfigv1alpha1.RecommendedDefaultServiceControllerConfiguration(&obj.ServiceController)
```
Comment (liggitt) [inline cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go:50]: I expected the defaulting functions for the service controller to accompany the service controller config type and be in `k8s.io/cloud-provider`

File: staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go [hunk 24]
```diff
@@ -18,8 +18,10 @@ package v1alpha1
 
 import (
 	"k8s.io/apimachinery/pkg/conversion"
+	cpconfig "k8s.io/cloud-provider/config"
+	serviceconfig "k8s.io/cloud-provider/service/config"
+	cmconfig "k8s.io/controller-manager/config"
 	"k8s.io/kube-controller-manager/config/v1alpha1"
```
Comment (liggitt) [inline staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:24]: the v1alpha1 GenericControllerManagerConfiguration type should move to k8s.io/controller-manager as well so this dependency is dropped

File: staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go [hunk 25]
```diff
@@ -18,8 +18,10 @@ package v1alpha1
 
 import (
 	"k8s.io/apimachinery/pkg/conversion"
+	cpconfig "k8s.io/cloud-provider/config"
+	serviceconfig "k8s.io/cloud-provider/service/config"
```
Comment (liggitt) [inline staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:22]: the v1alpha1 KubeCloudSharedConfiguration and ServiceControllerConfiguration types (and the conversions from this file) should move to `k8s.io/cloud-provider` to drop these dependencies

File: staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go [hunk 26]
```diff
@@ -18,8 +18,10 @@ package v1alpha1
```
Comment (liggitt) [inline staging/src/k8s.io/controller-manager/config/v1alpha1/conversion.go:18]: this file should be split up so that this repo only contains conversions for GenericControllerManagerConfiguration, and the rest of the conversions are moved to the cloud-provider repo

File: staging/src/k8s.io/controller-manager/go.mod (hunk 22 above)
Comment (cheftako) [reply staging/src/k8s.io/controller-manager/go.mod]: A little uncomfortable that we may have put too much (interface definition and controllers) into cloud-provider. Topic for another time.

File: cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go (hunk 23 above)
Comment (cheftako) [reply cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go]: +1

Comment (cici37) [reply cmd/cloud-controller-manager/app/apis/config/v1alpha1/defaults.go]: Updated.

**Comment by nikhita** [comment]
```
 F1008 07:46:48.473993     157 main.go:35] Invalid rules file "/home/prow/go/src/k8s.io/kubernetes/staging/publishing/rules.yaml": validation errors:
- repository "kube-controller-manager" cannot depend on "controller-manager" later in the rules file
- repository "kube-controller-manager" cannot depend on "cloud-provider" later in the rules file 
```

the `controller-manager` rules need to be moved above the `kube-controller-manager` rules in rules.yaml for the publishing-bot job to pass

**Comment by cici37** [comment]
@nikhita Thank you so much for the details 😊

**Comment by cici37** [comment]
Brief of this PR:

1. Moved `cmd/controller-manager` to `k8s.io/controller-manager`;
    - GenericControllerManagerConfiguration(from `pkg/controller/apis/config`) to `k8s.io/controller-manager/config`;

2. Moved `pkg/controller/service` to `k8s.io/cloud-provider/service` (ServiceControllerConfiguration);

3. Moved `cmd/cloud-controller-manager/app/apis/config` to `k8s.io/cloud-provider/app/apis/config`
    - CloudControllerManagerConfiguration(from `cmd/cloud-controller-manager/app/apis/config`), KubeCloudSharedConfiguration(from `pkg/controller/apis/config`) and CloudProviderConfiguration(from `pkg/controller/apis/config`) to `k8s.io/cloud-provider/app/apis/config`;

**Comment by cici37** [comment]
@liggitt @cheftako This PR is ready to review. Please have a look when have time😊

File: pkg/controller/apis/config/v1alpha1/doc.go [hunk 27]
```diff
@@ -17,6 +17,12 @@ limitations under the License.
 // +k8s:deepcopy-gen=package
 // +k8s:conversion-gen=k8s.io/kubernetes/pkg/controller/apis/config
 // +k8s:conversion-gen=k8s.io/component-base/config/v1alpha1
+// +k8s:conversion-gen=k8s.io/cloud-provider/app/apis/config
```
Comment (liggitt) [inline pkg/controller/apis/config/v1alpha1/doc.go:20]: drop the following conversion-gen references in a follow-up, we only need the external packages:
* `k8s.io/cloud-provider/app/apis/config`
* `k8s.io/cloud-provider/service/config`
* `k8s.io/controller-manager/config`

**Comment by liggitt** [comment]
/lgtm
/approve

one note for posterity, and a few requests for follow-up fixup

File: pkg/controller/apis/config/types.go [hunk 28]
```diff
@@ -18,7 +18,9 @@ package config
 
 import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	componentbaseconfig "k8s.io/component-base/config"
+	cpconfig "k8s.io/cloud-provider/app/apis/config"
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply pkg/controller/apis/config/types.go:21]: +1 to just `k8s.io/cloud-provider/config`

File: staging/src/k8s.io/csi-translation-lib/go.mod [hunk 29]
```diff
@@ -15,8 +15,10 @@ require (
 replace (
 	k8s.io/api => ../api
 	k8s.io/apimachinery => ../apimachinery
+	k8s.io/apiserver => ../apiserver
 	k8s.io/client-go => ../client-go
 	k8s.io/cloud-provider => ../cloud-provider
```
[the remark this answers is not in the document]
Comment (cici37) [reply staging/src/k8s.io/csi-translation-lib/go.mod:20]: https://github.com/kubernetes/kubernetes/issues/95530
